graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java

```diff
@@ -0,0 +1,161 @@
+    }
+
+    OperationDefinition operation = environment.getOperationDefinition();
+    if (operation.getDirectives().stream().anyMatch(d -> d.getName().equals(ATOMIC_DIRECTIVE))
+        && operation.getSelectionSet().getSelections().size() > 1) {
+      return Collections.singletonList(
+          executeAsBatch(environment, dataStore, queries, buildException));
+    }
+
+    if (buildException != null) {
+      CompletableFuture<Map<String, Object>> f = new CompletableFuture<>();
+      f.completeExceptionally(buildException);
+      return Collections.singletonList(f);
+    }
+
+    List<Map<String, Object>> values = environment.getArgument("values");
+    if (values.size() != queries.size()) {
+      throw new IllegalStateException("Number of values to insert should match number of queries");
+    }
+
+    List<CompletableFuture<Map<String, Object>>> results = new ArrayList<>(values.size());
+    for (int i = 0; i < queries.size(); i++) {
+      // Execute as a single statement
+      int finalI = i;
+      results.add(
+          dataStore
+              .execute(queries.get(i))
+              .thenApply(rs -> ImmutableMap.of("value", values.get(finalI))));
+    }
+    return results;
+  }
+
+  private CompletableFuture<Map<String, Object>> executeAsBatch(
+      DataFetchingEnvironment environment,
+      DataStore dataStore,
+      List<BoundQuery> queries,
+      Exception buildException) {
+    int sumOfAllQueriesForAllSelections = calculateNumberOfAllQueries(environment.getOperationDefinition());
+    HttpAwareContext context = environment.getContext();
+    HttpAwareContext.BatchContext batchContext = context.getBatchContext();
+
+    if (environment.getArgument("options") != null) {
+      // Users should specify query options once in the batch
+      boolean dataStoreAlreadySet = batchContext.setDataStore(dataStore);
+
+      if (dataStoreAlreadySet) {
+        // DataStore can be set at most once.
+        // The instance that should be used should contain the user options (if any).
+        buildException =
+            new GraphQLException(
+                "options can only de defined once in an @atomic mutation selection");
+      }
+    }
+
+    if (buildException != null) {
+      batchContext.setExecutionResult(buildException);
+    } else {
+      for (BoundQuery query : queries) {
+        if (batchContext.add(query) == sumOfAllQueriesForAllSelections) {
+          // All the statements were added successfully
+          // Use the dataStore containing the options
+          DataStore batchDataStore = batchContext.getDataStore().orElse(dataStore);
+          batchContext.setExecutionResult(batchDataStore.batch(batchContext.getQueries()));
+        }
+      }
+    }
+    return batchContext
+        .getExecutionFuture()
+        .thenApply(v -> ImmutableMap.of("values", environment.getArgument("values")));
+  }
+
+  private int calculateNumberOfAllQueries(OperationDefinition operationDefinition) {
```

Comment:
I think it would be easier to keep it based on the number of selections. You just need a separate variable in `BatchContext` because it's no longer equal to the number of queries.
```java
public static class BatchContext {
  private final List<BoundQuery> queries = new ArrayList<>();
  private final int operationCount;

  public synchronized int add(BoundQuery query) {
    queries.add(query);
    operationCount += 1;
    return operationCount;
  }

  public synchronized int add(List<BoundQuery> newQueries) {
    queries.add(newQueries);
    operationCount += newQueries.size();
    return operationCount;
  }
}
```

In `MutationFetcher.executeAsBatch`:
```java
    int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
...
    } else if (batchContext.add(query) == selections) {
      ... // execute the batch
    }
```

In `BulkMutationFetcher.executeAsBatch`:
```java
    int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
...
    } else if (batchContext.add(queries) == selections) {
      ... // execute the batch
    }
```

Reply:
This solution doesn't work. Let's consider a situation when we can have N selections, and each of the selections can have M queries. Let's assume that we have two selections. The first one has two queries, the second one has one query:
```
s1: [{q1}, {q2}]
s2: {q3}
```
for that case, the:
```
int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
```
evaluates to `2`. When the first selection is evaluated, and both queries are added, the 
```
public synchronized int add(List<BoundQuery> newQueries) {
```
will return two. So, this condition:
```
batchContext.add(queries) == selections
```
will be met and the batch will be executed. I am not sure if this should be executed yet, because the second selection contains one additional query. Shouldn't all 3 queries be batched and executed together as one query?

This is why I initially thought that the `BulkMutationFetcher` will execute all 3 queries, so then the selections must be equal to the total number of queries (3 for that case), and not selections. 
However, when debugging this I see that there will be two separate queries executed:
1. Batch query for `q1` and `a2` using the `BulkMutationFetcher`
2. Single query for `q3` using the `BulkMutationFetcher`
Is it really atomic if we have two separate queries? 
 
For this situation, the proposed solution is also problematic:
1. When the first selection with `q1` and `q2` arrives at `BulkMutationFetcher` it will be executed once both queries are added (because `2 == queries == selections`)
2. The `MutationFetcher` tries to execute the remaining `q3`.
It will be added to the `batchContext`, but after the `batchContext.add(query)` is executed, it returns `3` (it had 2 queries added by `BulkMutationFetcher` and the `q3`. However, the `selections == 2`, so the `q3` won't be executed at all. The problem is that the `BatchContext` is shared between both mutation fetchers.

The situation is getting more complicated if the first selection has more queries (i.e. 10 queries). For that case, the 
`BulkMutationFetcher#batchContext.add(queries) == selections)` won't be evaluated properly. 
Another problem when we switch the order of queries (first the selection with on query q1, then the selection with two queries). For this situation, not all bulk queries from the second selection will be executed.

Reply:
🤦 my bad, I meant to also increment by one when multiple queries are added:
```java
public static class BatchContext {
  private final List<BoundQuery> queries = new ArrayList<>();
  private final int operationCount;

  public synchronized int add(BoundQuery query) {
    queries.add(query);
    operationCount += 1;
    return operationCount;
  }

  public synchronized int add(List<BoundQuery> newQueries) {
    queries.add(newQueries);
    operationCount += 1;
    return operationCount;
  }
}
```

Reply:
Yeah, your solution looks simpler. I've implemented it in:
https://github.com/stargate/stargate/pull/874/commits/c358b1012c7b0eec0183f877a3bde8bb3d391059